app/models/user_auth_token.rb

```diff
@@ -30,6 +30,40 @@ def self.log(info)
     end
   end
 
+  # Returns the login location as it will be used by the the system to detect
+  # suspicious login.
+  #
+  # This should not be very specific because small variations in location
+  # (i.e. changes of network, small trips, etc) will be detected as suspicious
+  # logins.
+  #
+  # On the other hand, if this is too broad it will not report any suspicious
+  # logins at all.
+  #
+  # For example, let's choose the country as the only component in login
+  # locations. In general, this should be a pretty good choce with the
+  # exception that for users from huge countries it might not be specific
+  # enoguh. For US users where the real user and the malicious one could
+  # happen to live both in USA, this will not detect any suspicious activity.
+  def self.login_location(ip)
+    DiscourseIpInfo.get(ip)[:country]
+  end
+
+  def self.is_suspicious(user_id, user_ip)
+    ips = UserAuthTokenLog.where(user_id: user_id).pluck('DISTINCT client_ip')
+    return false if ips.empty? # first login is never suspicious
```

Comment:
Is this ever hit? Looks like `is_suspicious` is called **after** you've [created](https://github.com/discourse/discourse/blob/651d911a4cfa8beb1975b2e97ae239403ce5cbc8/app/models/user_auth_token.rb#L70-L77) the `UserAuthToken` for the current session

Reply:
Yes, this checks the UserAuthToken**Log**s and is called before log is called.

There is a test for this one.

Reply:
My bad. I apparently can't read 🙈